src/metabase/models/field_values.clj

```diff
@@ -106,36 +106,50 @@
             (#{:list :auto-list} (keyword has-field-values)))))))
 
-(defn- values-less-than-total-max-length?
-  "`true` if the combined length of all the values in [[distinct-values]] is below the threshold for what we'll allow in a
-  FieldValues entry. Does some logging as well."
-  [distinct-values]
-  ;; only consume enough values to determine whether the total length is > `total-max-length` -- if it is, we can stop
-  (let [total-length (reduce
-                      (fn [total-length v]
-                        (let [new-total (+ total-length (count (str v)))]
-                          (if (>= new-total total-max-length)
-                            (reduced new-total)
-                            new-total)))
-                      0
-                      distinct-values)]
-    (u/prog1 (<= total-length total-max-length)
-      (log/debug (trs "Field values total length is > {0}." total-max-length)
-                 (if <>
-                   (trs "FieldValues are allowed for this Field.")
-                   (trs "FieldValues are NOT allowed for this Field."))))))
+
+(defn take-by-length
+  "Returns a transducer that takes a collection and returns a collection where the total length of all items in list is less than `max-length`."
+  [max-length]
+  (fn [rf]
+    (let [current-length (volatile! 0)]
+      (fn
+        ([] (rf))
+        ([result]
+         (rf result))
+        ([result input]
+         (vswap! current-length + (count (str input)))
+         (if (< @current-length max-length)
+           (rf result input)
+           (reduced result)))))))
 
 (defn distinct-values
   "Fetch a sequence of distinct values for `field` that are below the [[total-max-length]] threshold. If the values are
-  past the threshold, this returns `nil`. (This function provides the values that normally get saved as a Field's
+  past the threshold, this returns a subset of possible values values where the total length of all items is less than [[total-max-length]].
+  It also returns a `has_more_values` flag, `has_more_values` = `true` when the returned values list is a subset of all possible values.
+
+  ;; (distinct-values (Field 1))
+  ;; ->  {:values          [1, 2, 3]
+          :has_more_values false}
+
+  (This function provides the values that normally get saved as a Field's
   FieldValues. You most likely should not be using this directly in code outside of this namespace, unless it's for a
   very specific reason, such as certain cases where we fetch ad-hoc FieldValues for GTAP-filtered Fields.)"
   [field]
   (classloader/require 'metabase.db.metadata-queries)
   (try
-    (let [values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)]
-      (when (values-less-than-total-max-length? values)
-        values))
+    (let [distinct-values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)
+          values          (into [] (take-by-length total-max-length) distinct-values)]
```

Comment:
Maybe call this `limited-values` or `limited-distinct-values` or something? `values` feels like it should be a superset of `distinct-values` instead of a subset, but maybe that's just me

Reply:
I agree, `limited-distinct-values` makes much more sense